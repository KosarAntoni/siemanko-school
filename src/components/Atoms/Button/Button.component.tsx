import React from 'react';
import clsx from 'clsx';
import Link from 'next/link';

import styles from './Button.styles.module.scss';

import { ButtonProps } from './Button.models';

const Button = ({
  children,
  className,
  url,
  onClick,
  color = 'Orange',
  style = 'Solid',
  size = 'Regular',
  type = 'button',
}: ButtonProps): JSX.Element => {
  const buttonClasses = clsx(
    styles.button,
    color && styles[`button${color}`],
    style && styles[`button${style}`],
    size && styles[`button${size}`],
    className,
  );

  if (url) {
    const internal = /^\/(?!\/)/.test(url);

    return internal ? (
      <Link href={url} className={buttonClasses}>
        {children}
      </Link>
    ) : (
      <a href={url} className={buttonClasses} target="_blank">
        {children}
      </a>
    );
  }

  if (onClick) {
    return (
      <button type={type} onClick={onClick} className={buttonClasses}>
        {children}
      </button>
    );
  }

  return (
    <span className={buttonClasses} data-testid="button-item">
      {children}
    </span>
  );
};

export default Button;
