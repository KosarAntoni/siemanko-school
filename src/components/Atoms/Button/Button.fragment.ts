export const buttonFragment = `fragment ButtonFragment on Button {
  text
  url
  reference {
    slug
  }
  style
  color
  size
}`;
