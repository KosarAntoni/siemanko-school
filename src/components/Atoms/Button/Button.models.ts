import { ButtonHTMLAttributes, PropsWithChildren } from 'react';

export interface ButtonProps extends PropsWithChildren, Pick<ButtonHTMLAttributes<HTMLButtonElement>, 'type'> {
  className?: string;
  url?: string | null;
  onClick?: () => void;
  color?: 'Orange' | 'White';
  style?: 'Solid' | 'Outline';
  size?: 'Big' | 'Regular' | 'Small';
  text?: string;
}
