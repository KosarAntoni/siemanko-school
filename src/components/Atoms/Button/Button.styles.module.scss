.button {
  padding: 1.2rem 1.5rem;
  color: var(--white);
  font-weight: 600;
  font-size: 1.6rem;
  background-color: var(--orange);
  border-color: var(--orange);
  border-style: solid;
  border-radius: 0.5rem;
  transition: transform 0.3s;

  &Solid {
    &:hover {
      transform: translateY(-1px);
    }
  }

  &White {
    color: var(--orange);
    background-color: var(--white);
    border-color: var(--white);
  }

  &Outline {
    background-color: transparent;
    transition: background-color 0.3s, color 0.3s;

    &.buttonWhite {
      color: var(--white);

      &:hover {
        color: var(--orange);
        background-color: var(--white);
      }
    }

    &.buttonOrange {
      color: var(--orange);

      &:hover {
        color: var(--white);
        background-color: var(--orange);
      }
    }
  }

  &Big {
    padding: 1.5rem 2rem;
  }

  &Small {
    padding: 0.5rem 1.5rem;
    font-size: 1.3rem;
  }
}

button.button {
  cursor: pointer;
}

a.button {
  display: inline-block;
}
