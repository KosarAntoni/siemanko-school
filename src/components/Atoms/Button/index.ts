export { default } from './Button.component';
export * from './Button.models';
export * from './Button.fragment';
