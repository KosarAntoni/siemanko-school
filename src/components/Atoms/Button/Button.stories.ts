import type { Meta, StoryObj } from '@storybook/react';

import Button from '.';

const meta: Meta<typeof Button> = {
  title: 'Atoms/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    color: {
      control: 'radio',
      options: ['Orange', 'White'],
    },
    size: {
      control: 'radio',
      options: ['Small', 'Regular', 'Big'],
    },
    style: {
      control: 'radio',
      options: ['Solid', 'Outline'],
    },
    children: {
      control: 'text',
      name: 'text',
    },
  },
  args: {
    color: 'Orange',
    size: 'Regular',
    style: 'Solid',
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const AsSpan: Story = {
  args: {
    children: 'Button',
  },
};

export const AsLink: Story = {
  args: {
    children: 'Button',
    url: '/test',
  },
};

export const AsButton: Story = {
  args: {
    children: 'Button',
    onClick: () => {
      alert('Click');
    },
  },
};
