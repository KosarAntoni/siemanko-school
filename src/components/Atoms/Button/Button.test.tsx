import { fireEvent, render, screen } from '@testing-library/react';

import Button from '.';

describe('Button', () => {
  it('renders a Button as button', () => {
    const testFunc = jest.fn();
    render(<Button onClick={testFunc}>Test</Button>);

    const buttonNode = screen.getByRole('button', {
      name: 'Test',
    });

    expect(buttonNode).toMatchSnapshot();

    expect(buttonNode).toBeInTheDocument();

    fireEvent.click(buttonNode);
    expect(testFunc).toHaveBeenCalled();

    expect(buttonNode).toHaveClass('button buttonOrange buttonSolid buttonRegular');
  });

  it('renders a Button as external link', () => {
    const url = 'https://www.test.com/';
    render(
      <Button url={url} style="Outline" color="White" size="Small">
        Test
      </Button>,
    );

    const buttonNode = screen.getByRole('link', {
      name: 'Test',
    });

    expect(buttonNode).toMatchSnapshot();

    expect(buttonNode).toBeInTheDocument();

    expect(buttonNode).toHaveAttribute('href', url);

    expect(buttonNode).toHaveAttribute('target', '_blank');

    expect(buttonNode).toHaveClass('button buttonWhite buttonOutline buttonSmall');
  });

  it('renders a Button as intenal link', () => {
    const url = '/test';
    render(<Button url={url}>Test</Button>);

    const buttonNode = screen.getByRole('link', {
      name: 'Test',
    });

    expect(buttonNode).toMatchSnapshot();

    expect(buttonNode).toBeInTheDocument();

    expect(buttonNode).toHaveAttribute('href', url);
  });

  it('renders a Button as span', () => {
    render(<Button size="Big">Test</Button>);

    const buttonNode = screen.getByText('Test');

    expect(buttonNode).toMatchSnapshot();

    expect(buttonNode).toBeInTheDocument();

    expect(buttonNode).toHaveClass('button buttonOrange buttonSolid buttonBig');
  });
});
