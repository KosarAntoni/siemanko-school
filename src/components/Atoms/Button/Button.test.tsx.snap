// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly renders Button as button 1`] = `
<button
  class="button buttonOrange buttonSolid buttonRegular"
  type="button"
>
  Test
</button>
`;

exports[`Button component renders correctly renders Button as external link 1`] = `
<a
  class="button buttonWhite buttonOutline buttonSmall"
  href="https://www.test.com/"
  target="_blank"
>
  Test
</a>
`;

exports[`Button component renders correctly renders Button as intenal link 1`] = `
<a
  class="button buttonOrange buttonSolid buttonRegular"
  href="/test"
>
  Test
</a>
`;

exports[`Button component renders correctly renders a Button as button 1`] = `
<button
  class="button buttonOrange buttonSolid buttonRegular"
  type="button"
>
  Test
</button>
`;

exports[`Button component renders correctly renders a Button as external link 1`] = `
<a
  class="button buttonWhite buttonOutline buttonSmall"
  href="https://www.test.com/"
  target="_blank"
>
  Test
</a>
`;

exports[`Button component renders correctly renders a Button as intenal link 1`] = `
<a
  class="button buttonOrange buttonSolid buttonRegular"
  href="/test"
>
  Test
</a>
`;

exports[`Button component renders correctly renders a Button as span 1`] = `
<span
  class="button buttonOrange buttonSolid buttonBig"
  data-testid="button-item"
>
  Test
</span>
`;

exports[`Button renders a Button as button 1`] = `
<button
  class="button buttonOrange buttonSolid buttonRegular"
  type="button"
>
  Test
</button>
`;

exports[`Button renders a Button as external link 1`] = `
<a
  class="button buttonWhite buttonOutline buttonSmall"
  href="https://www.test.com/"
  target="_blank"
>
  Test
</a>
`;

exports[`Button renders a Button as intenal link 1`] = `
<a
  class="button buttonOrange buttonSolid buttonRegular"
  href="/test"
>
  Test
</a>
`;

exports[`Button renders a Button as span 1`] = `
<span
  class="button buttonOrange buttonSolid buttonBig"
  data-testid="button-item"
>
  Test
</span>
`;

exports[`a Button component renders correctly renders a Button as button 1`] = `
<button
  class="button buttonOrange buttonSolid buttonRegular"
  type="button"
>
  Test
</button>
`;

exports[`a Button component renders correctly renders a Button as external link 1`] = `
<a
  class="button buttonWhite buttonOutline buttonSmall"
  href="https://www.test.com/"
  target="_blank"
>
  Test
</a>
`;

exports[`a Button component renders correctly renders a Button as intenal link 1`] = `
<a
  class="button buttonOrange buttonSolid buttonRegular"
  href="/test"
>
  Test
</a>
`;

exports[`a Button component renders correctly renders a Button as span 1`] = `
<span
  class="button buttonOrange buttonSolid buttonBig"
  data-testid="button-item"
>
  Test
</span>
`;

exports[`the Button component renders correctly renders a Button as button 1`] = `
<button
  class="button buttonOrange buttonSolid buttonRegular"
  type="button"
>
  Test
</button>
`;

exports[`the Button component renders correctly renders a Button as external link 1`] = `
<a
  class="button buttonWhite buttonOutline buttonSmall"
  href="https://www.test.com/"
  target="_blank"
>
  Test
</a>
`;

exports[`the Button component renders correctly renders a Button as intenal link 1`] = `
<a
  class="button buttonOrange buttonSolid buttonRegular"
  href="/test"
>
  Test
</a>
`;

exports[`the Button component renders correctly renders a Button as span 1`] = `
<span
  class="button buttonOrange buttonSolid buttonBig"
  data-testid="button-item"
>
  Test
</span>
`;
